cmd/kube-controller-manager/controller-manager.go

```diff
@@ -46,10 +65,32 @@ func main() {
 	util.InitLogs()
 	defer util.FlushLogs()
 
-	verflag.PrintAndExitIfRequested()
+	//HA Function, this function will be called by the lease manager itself.
+	startRC := func() bool {
+		RUNNING = true
+		if err := s.Run(pflag.CommandLine.Args()); err != nil {
+			fmt.Fprintf(os.Stderr, "%v\n", err)
+			glog.Infof("EXITING NOW ! Killed")
+			os.Exit(1)
+		}
+		return true
+	}
 
-	if err := s.Run(pflag.CommandLine.Args()); err != nil {
-		fmt.Fprintf(os.Stderr, "%v\n", err)
-		os.Exit(1)
+	//Acquire a lock before starting.
+	mcfg := controller.Config{
+		EtcdServers: "http://localhost:4001",
+		Key:         "rcLEASE",
+		Running:     isRunning,
+		Lease:       startRC,
+		Unlease:     endRC}
```

Comment:
May want to call it FailedLeaseRenewal vs. Unlease ;-) which is not a word/verb.